### PR TITLE
Convert resource_from_hash to a custom resource and simplify

### PR DESCRIPTION
**Pull request by jcorme**
I simplified the resource definition into something functionally identical, as far as I can tell. The original solution of generating a string and using eval seems unnecessarily hack-ish... I've tested these changes on every cookbook that uses resource_from_has in a recipe, and they all pass (except for osl-haproxy which has unrelated data bag encryption key errors on two of the suites).

Switching to a custom resource, hash apparently can't be used as a property name, so we'll have to update that in base, osl-mirror, and osl-haproxy.

These changes are primarily to fix using custom resources in Chef 13, since Chef no longer creates a Chef::Resources::<ResourceName> constant for any files under the resources directory.

### CODE REVIEW

**Comment by ramereth** [comment]
[Chef 13: resource_from_hash](https://trello.com/c/Dr53ohLW/944-chef-13-resourcefromhash)

File: resources/resource_from_hash.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+action :do do
+  attrs = indifferent_access(attrs_hash, 'attributes')
+  res_type = indifferent_access(attrs_hash, 'resource').to_sym
+  res_name = indifferent_access(attrs_hash, 'name')
+
+  declare_resource(res_type, res_name) do
+    attrs.to_a.each do |attr|
+      key, val = attr
+      self.send(key.to_sym, val)
```
Comment (ramereth) [inline resources/resource_from_hash.rb]: rubocop complains about this

File: test/integration/cookbooks/resource_from_hash_test/recipes/default.rb [hunk 2]
```diff
@@ -26,14 +26,14 @@
 }
 
 resource_from_hash 'test' do
-  hash data
+  attrs_hash data
   action :do
```
Comment (ramereth) [inline test/integration/cookbooks/resource_from_hash_test/recipes/default.rb]: This is the default action, so please remove it here and below.

File: resources/resource_from_hash.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    end
+  end
+end
+
+action :log do
+  Chef::Log.info "Logging resource hash #{name} "\
+    "with the following attributes: #{attrs_hash}"
```
Comment (ramereth) [inline resources/resource_from_hash.rb]: Make this one line please.

**Comment by ramereth** [comment]
@jcorme btw, once this is merged, let's make a separate PR to do the Chef 13 fixes

**Comment by ramereth** [comment]
@jcorme the LGTM however I'd like to see PRs from those three cookbooks which resolve the new issue so that we can make this change all at once.

**Comment by ramereth** [comment]
@jcorme looks like you already made the change in osuosl-cookbooks/osl-mirror#74. I'd prefer to make a separate PR just for that so we're not waiting on that larger PR to be ready.

**Comment by jcorme** [comment]
@ramereth PRs created.

**Comment by jcorme** [comment]
!bump major *

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a major-level version bump to v1.0.0.  Have a nice day!

**Pull request merged**